jso/apis/src/main/java/org/teavm/jso/core/JSPromise.java

```diff
@@ -0,0 +1,103 @@
+/*
+ *  Copyright 2023 Bernd Busse.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.teavm.jso.core;
+
+import org.teavm.interop.NoSideEffects;
+import org.teavm.jso.JSBody;
+import org.teavm.jso.JSByRef;
+import org.teavm.jso.JSFunctor;
+import org.teavm.jso.JSMethod;
+import org.teavm.jso.JSObject;
+
+import org.teavm.jso.util.function.JSConsumer;
+import org.teavm.jso.util.function.JSFunction;
+import org.teavm.jso.util.function.JSSupplier;
+
+/**
+ * Interface for interacting with JavaScript <a href="https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise">{@code Promise}s</a>.
+ *
+ * @param <T> The type this promise returns when resolving successfully.
+ * @param <E> The type this promise returns when rejecting.
+ */
+public abstract class JSPromise<T, E> implements JSObject {
+    private JSPromise() {
+    }
+
+    /** Interface for a function wrapped by a promise. */
+    @FunctionalInterface
+    @JSFunctor
+    public interface Executor<T, E> extends JSObject {
+        /**
+         * @param resolveFunc Call this function to resolve with success value.
+         * @param rejectFunc Call this function to reject with error value.
+         */
+        void onExecute(JSConsumer<T> resolveFunc, JSConsumer<E> rejectFunc);
+    }
+
+    @JSBody(params = "executor", script = "return new Promise(executor);")
+    @NoSideEffects
+    public static native <T, E> JSPromise<T, E> create(Executor<T, E> executor);
+
+    @JSBody(params = "promises", script = "return Promise.any(promises);")
+    @NoSideEffects
+    public static native JSPromise<?, ? extends JSError> any(@JSByRef JSArray promises);
```

Comment:
1. This method should be generic
2. Instead of `JSArray<T>`, parameter should be `JSArrayReader<T>`. Actually, according to spec it must be `JSIterable<T>`, however, TeaVM does not support symbols and can't declare `JSIterable<T>`, so for now `JSArrayReader<T>` must be enough.

Reply:
The `@JSByRef` annotation is also wrong...